Archipelago.HollowKnight/HintTracker.cs

```diff
@@ -24,35 +25,63 @@ public static class HintTracker
     private static void UpdateHints(Hint[] arrayHints)
     {
         Hints = arrayHints.ToList();
-        foreach (var hint in Hints)
+        foreach (Hint hint in Hints)
         {
             if (hint.FindingPlayer != Archipelago.Instance.session.ConnectionInfo.Slot)
                 continue;
 
-            var locationName = StripShopSuffix(_session.Locations.GetLocationNameFromId(hint.LocationId));
-            var location = Finder.GetLocation(locationName);
-            if (location == null)
-                continue;
-            if (!Ref.Settings.Placements.ContainsKey(locationName))
-                continue;
-
-            var placement = Ref.Settings.Placements[locationName];
+            if (!Archipelago.Instance.placementsByLocationID.ContainsKey(hint.LocationId)) continue;
+            AbstractPlacement placement = Archipelago.Instance.placementsByLocationID[hint.LocationId];
 
             if (placement == null)
                 continue;
-            
-            //get all items in the placement and set the hinted attribute in ArchipelagoItemTag to true.
-            foreach (var tag in placement.Items.Select(item => item.GetTag<ArchipelagoItemTag>()).Where(tag => tag.Location == hint.LocationId))
+
+            // set the hinted tag for the single item in the placement that was hinted for.
+            foreach (ArchipelagoItemTag tag in placement.Items.Select(item => item.GetTag<ArchipelagoItemTag>())
+                         .Where(tag => tag.Location == hint.LocationId))
             {
                 tag.Hinted = true;
             }
-            
-            //if there is only 1 item in this placement mark the whole placement as visited.
-            if (placement.Items.Count > 1) continue;
-            placement.GetTag<ArchipelagoPlacementTag>().Hinted = true;
-            placement.AddVisitFlag(VisitState.Previewed);
+
+            // if all items inside a placement have been hinted for then mark the entire placement as hinted.
+            if (placement.Items.TrueForAll(item => item.GetTag<ArchipelagoItemTag>().Hinted))
+            {
+                placement.GetTag<ArchipelagoPlacementTag>().Hinted = true;
+            }
+
+            if (placement is ShopPlacement shop)
+            {
+                List<(string, AbstractItem)> previewText = new();
+                foreach (AbstractItem item in shop.Items)
+                {
+                    if (item.GetTag<ArchipelagoItemTag>().Hinted)
+                    {
+                        previewText.Add((item.GetPreviewWithCost(), item));
+                    }
+                    else
+                    {
+                        previewText.Add((Language.Language.Get("???", "IC"), item));
+                    }
+
+                }
+
+                shop.OnPreviewBatch(previewText);
+            }
+            else
+            {
+                List<string> previewText = new();
```

Comment:
Does placement.getuiname not do this whole sequence?

Reply:
not with the cost.

Reply:
Makes sense